File: hs-src/lib/Tokenizer.hs
{-# LANGUAGE DeriveDataTypeable #-}

module Tokenizer
       where

import Data.Typeable  -- used by our custom Test.Hasty node testrunner
import Data.Char

data Operator = Plus | Minus | Times | Div | GreaterThan | LessThan |
                Equals | Not | Mod
  deriving (Show, Read, Eq, Typeable)

data Token = TokenEquals            |
             TokenOperator Operator |
             TokenNumber   Double   |
             TokenSymbol   Char     |
             TokenComma             |
             TokenLeftParen         |
             TokenRightParen        |
             TokenLeftBrace         |
             TokenRightBrace        |
             TokenDefun             |
             TokenIf                |
             TokenRepeat
           deriving (Show, Read, Eq, Typeable)

tokenize :: String -> [Token]
tokenize [] = []
tokenize (c : cs)
  | isSpace c         = tokenize cs
  | isAlpha c         = TokenSymbol c : tokenize cs
  | c == ','          = TokenComma : tokenize cs
  | c == '='          = TokenEquals     : tokenize cs
  | c == '('          = TokenLeftParen  : tokenize cs
  | c == ')'          = TokenRightParen : tokenize cs
  | c == '{'          = TokenLeftBrace  : tokenize cs
  | c == '}'          = TokenRightBrace : tokenize cs
  | c == '~'          = TokenOperator Equals : tokenize cs
  | c == '+'          = TokenOperator Plus  : tokenize cs
  | c == '-'          = TokenOperator Minus : tokenize cs
  | c == '*'          = TokenOperator Times : tokenize cs
  | c == '/'          = TokenOperator Div   : tokenize cs
  | c == '\\'         = TokenOperator Mod   : tokenize cs
  | c == '!'          = TokenOperator Not   : tokenize cs
  | c == '>'          = TokenOperator GreaterThan : tokenize cs
  | c == '<'          = TokenOperator LessThan    : tokenize cs
  | c == '&'          = TokenDefun            : tokenize cs
  | c == '?'          = TokenIf               : tokenize cs
  | c == '#'          = TokenRepeat           : tokenize cs
  | isDigit c         = let (numstr, cs') = whileNum (c:cs) ""
                        in TokenNumber (read numstr) : tokenize cs'
  | otherwise         = error $ "could not tokenize " ++ [c]
  where whileNum (c : cs) acc
          | isDigit c                         = whileNum cs (acc ++ [c])
          | c == '.' && (not $ elem '.' acc)  = whileNum cs (acc ++ [c])
        whileNum cs acc = (acc, cs)

File: js-src/Tokenizer.js

exports.tokenize = tokenize


var TokenCases = [
  // isSpace case is handled by a trim down in the driver function
  [ /^[a-zA-Z\^\%\'\"]/, function (m) { return TokenSymbol (m[0]) } ],
  [ ',', TokenComma ],
  [ '=', TokenEquals ],
  [ '(', TokenLeftParen ],
  [ ')', TokenRightParen ],
  [ '{', TokenLeftBrace ],
  [ '}', TokenRightBrace ],
  [ '+', TokenOperator, 'Plus' ],
  [ '-', TokenOperator, 'Minus' ],
  [ '*', TokenOperator, 'Times' ],
  [ '/', TokenOperator, 'Div' ],
  [ '\\', TokenOperator, 'Mod' ],
  [ '~', TokenOperator, 'Equals' ],
  [ '!', TokenOperator, 'Not' ],
  [ '>', TokenOperator, 'GreaterThan' ],
  [ '<', TokenOperator, 'LessThan' ],
  [ '&', TokenDefun ],
  [ '?', TokenIf ],
  [ '#', TokenRepeat ],
  [ /^[0-9]+(\.[0-9]+)?/,   function (m) { return TokenNumber (+m[0]) } ]
]


function tokenize (stream) {
  var r = {}
  stream = stream.replace (/^\s+/,'')
  if (! stream) {
    return []
  }
  r.stream = stream
  if (! TokenCases.some (tryCase, r)) {
    throw "could not tokenize: " + stream
  }

  
  return [r.result.token].concat(tokenize(r.result.stream))
}


function tryCase (tcase) {
  if (r = tryLiteral(this.stream,tcase) || tryRegExp(this.stream,tcase)) {
    this.result = r
    return true
  }
  return false
}

function tryLiteral (stream, tcase) {
  var cas = tcase[0]
  if (stream.indexOf(cas) == 0) {  // duck type, here (test only
                                   // succeeds if cas is a string)
    return { token: tcase[1](tcase.slice(2)),
             stream: stream.substr(cas.length) }
  }
  return false;
}

function tryRegExp (stream, tcase) {
  var cas = tcase[0]
  var m
  if (cas.test && (m=stream.match(cas))) {
    var matched_chars = m[0]
    var token = tcase[1](m)
    return { token: token, stream: stream.substr(matched_chars.length) }
  }
  return false;
}

//

var Token = {
  type:    null,
  is:      function (s) { return s === this.type },
  inspect: function () { return (this.type + (this.v ? (' ' + this.v) : '')) }
}

function TokenSymbol (c) {
  return Object.create (Token, { type: {value: 'TokenSymbol'},
                                 v:    {value: c },
                                 inspect: { value:
    function () { 
      return (this.type + (this.v ? (" '" + this.v + "'") : ''))
    }}})
}

function TokenNumber (num) {
  return Object.create (Token, { type: {value: 'TokenNumber'},
                                 v:    {value: num } })
}

function TokenComma (c) {
  return Object.create (Token, { type: {value: 'TokenComma'} })
}

function TokenEquals (c) {
  return Object.create (Token, { type: {value: 'TokenEquals'} })
}

function TokenLeftParen () {
  return Object.create (Token, { type: {value: 'TokenLeftParen'} })
}

function TokenRightParen () {
  return Object.create (Token, { type: {value: 'TokenRightParen'} })
}

function TokenLeftBrace () {
  return Object.create (Token, { type: {value: 'TokenLeftBrace'} })
}

function TokenRightBrace () {
  return Object.create (Token, { type: {value: 'TokenRightBrace'} })
}

function TokenOperator (op) {
  return Object.create (Token, { type: {value: 'TokenOperator'},
                                 v: {value: op} })
}

function TokenDefun (c) {
  return Object.create (Token, { type: {value: 'TokenDefun'} })
}

function TokenIf (c) {
  return Object.create (Token, { type: {value: 'TokenIf'} })
}

function TokenRepeat (c) {
  return Object.create (Token, { type: {value: 'TokenRepeat'} })
}
 

